pubtools/pulplib/_impl/client/ud_mappings.py

```diff
@@ -164,17 +178,23 @@ def upload_changed_mappings(mappings, repo, repo_url, do_request):
 
 
 def update_mappings_for_files(mappings, file_page):
-    # Updates mappings for every file in a single page, plus all
-    # following pages (async).
+    # Add missing mappings for every file in a single page, plus all
+    # following pages (async). Once all pages have been processed,
+    # obsolete mappings are removed.
     #
     # Returns Future[mappings] once all pages are processed.
 
     for unit in file_page.data:
         version = unit.version
         if version:
+            # Add missing mappings
             mappings.set_file_mapping(version, unit.path, unit.display_order)
+            # Save all the files for which the mapping should exist
+            mappings.filenames_from_pulp.append(unit.path)
```

Comment:
I don't think this statement should have been under the conditional.  There are a lot of files in practice which don't have 'version' set and yet shouldn't be removed from the mappings.